src/test/java/io/debezium/connector/vitess/VitessConnectorIT.java

```diff
@@ -836,6 +836,129 @@ public void testCopyAndReplicateTable() throws Exception {
         assertInsert(INSERT_NUMERIC_TYPES_STMT, schemasAndValuesForNumericTypes(), TestHelper.PK_FIELD);
     }
 
+    @Test
+    public void testVgtidIncludesLastPkDuringTableCopy() throws Exception {
+        TestHelper.executeDDL("vitess_create_tables.ddl");
+        int expectedSnapshotRecordsCount = 10;
+        final String tableName = "numeric_table";
+        for (int i = 1; i <= expectedSnapshotRecordsCount; i++) {
+            TestHelper.execute(INSERT_NUMERIC_TYPES_STMT, TEST_UNSHARDED_KEYSPACE);
+        }
+        String tableInclude = TEST_UNSHARDED_KEYSPACE + "." + tableName + "," + TEST_UNSHARDED_KEYSPACE + "." + tableName;
+        startConnector(Function.identity(), false, false, 1,
+                -1, -1, tableInclude, VitessConnectorConfig.SnapshotMode.INITIAL, TestHelper.TEST_SHARD);
+
+        // We should receive a record written before starting the connector.
+        consumer = testConsumer(expectedSnapshotRecordsCount);
+        consumer.await(TestHelper.waitTimeForRecords(), TimeUnit.SECONDS);
+        for (int i = 1; i <= expectedSnapshotRecordsCount; i++) {
+            SourceRecord record = assertRecordInserted(topicNameFromInsertStmt(INSERT_NUMERIC_TYPES_STMT), TestHelper.PK_FIELD);
+            assertSourceInfo(record, TEST_SERVER, TEST_UNSHARDED_KEYSPACE, tableName);
+            assertRecordSchemaAndValues(schemasAndValuesForNumericTypes(), record, Envelope.FieldName.AFTER);
+
+            if (i == expectedSnapshotRecordsCount) {
+                Map<String, ?> prevOffset = record.sourceOffset();
+                Map<String, ?> prevPartition = record.sourcePartition();
+                Testing.print(String.format("Offset: %s, partition: %s", prevOffset, prevPartition));
+                final String vgtidStr = (String) prevOffset.get(SourceInfo.VGTID_KEY);
+                final String expectedJSONString = "[{\"keyspace\":\"test_unsharded_keyspace\",\"shard\":\"0\"," +
+                        "\"gtid\":\"MySQL56/6a18875e-6d37-11ee-ac9a-0242ac110002:1-224\"," +
+                        "\"table_p_ks\":[{\"table_name\":\"numeric_table\",\"lastpk\":" +
+                        "{\"fields\":[{\"name\":\"id\",\"type\":\"INT64\",\"charset\":63,\"flags\":49667}]," +
+                        "\"rows\":[{\"lengths\":[\"2\"],\"values\":\"10\"}]}}]}]";
+                Vgtid actualVgtid = Vgtid.of(vgtidStr);
+                Vgtid expectedVgtid = Vgtid.of(expectedJSONString);
+                assertThat(actualVgtid.getShardGtids().size()).isEqualTo(1);
+                assertThat(actualVgtid.getShardGtids().get(0).getTableLastPrimaryKeys()).isEqualTo(
+                        expectedVgtid.getShardGtids().get(0).getTableLastPrimaryKeys());
+            }
+        }
+    }
+
+    @Test
+    public void testSnapshotLargeTable() throws Exception {
+        TestHelper.executeDDL("vitess_create_tables.ddl");
```

Comment:
@twthorn Could you please store all the records captured and assert at the end that all are present and there are neither gaps nor duplicates?